### PR TITLE
docs: update state management documentation

### PR DESCRIPTION
**Pull request by maartenbreddels**

### CODE REVIEW

File: solara/website/pages/documentation/getting_started/content/05-fundamentals/50-state-management.md [hunk 1]
```diff
@@ -33,21 +33,26 @@ def Page():
...
+allow you to create multiple instances of the same component with different states. But if the state reflects application state, there is by definition only one instance
+of it needed.
 
-[`solara.use_state`](/documentation/api/hooks/use_state) is a hook that allows you to manage local state within a specific component. This approach is beneficial when you want to encapsulate state within a component, making it self-contained and modular. Local state management is suitable for situations where state changes only affect the component and do not need to be shared across the application.
+### Local component state using solara.use_reactive
 
-Example:
-```solara
+If you do need state that is specific to a component, you should use [`solara.use_reactive`](/documentation/api/hooks/use_reactive) hook. These hooks allow you to create local state variables that are scoped to a specific component. This approach is useful when you want to encapsulate state within a component, making it self-contained and modular. Local state management is suitable for situations where state changes only affect the component and do not need to be shared across the application.
```
Comment (iisakkirotko) [inline solara/website/pages/documentation/getting_started/content/05-fundamentals/50-state-management.md:42]:
```suggestion
If you do need state that is specific to a component, you should use [`solara.use_reactive`](/documentation/api/hooks/use_reactive) hook. This hook allows you to create local state variables that are scoped to a specific component. This approach is useful when you want to encapsulate state within a component, making it self-contained and modular. Local state management is suitable for situations where state changes only affect the component and do not need to be shared across the application.
```

File: solara/website/pages/documentation/getting_started/content/05-fundamentals/50-state-management.md [hunk 2]
```diff
@@ -57,22 +62,27 @@ def Page():
...
 
-## Local component state using solara.use_reactive
+### Local component state using solara.use_state (not recommended)
 
+[`solara.use_state`](/documentation/api/hooks/use_state) is a hook that might be a bit more familiar to React developers. It also allows you to create local state variables that are scoped to a specific component, however, instead of using reactive variables, it uses a tuple of a value and a setter function.
 
-`use_reactive` is the middle ground between `use_state` and `reactive`. It allows you to create a reactive variable that is scoped to a specific component. This is more a matter of taste, we generally recommend using `use_reactive`, but if you prefer a little less magic, you can use `use_state` instead.
+We generally recommend using `use_reactive` over `use_state` as it is more easy to refactor between global application state and local component state by switching between `use_reactive` and `reactive`. There is no equivalent for `use_state` at the global level.
 
+If we take the previous example and replace `use_reactive` by `use_reactive`, we get:
```
Comment (iisakkirotko) [inline solara/website/pages/documentation/getting_started/content/05-fundamentals/50-state-management.md:71]:
```suggestion
If we take the previous example and replace `use_reactive` by `use_state`, we get:
```

File: solara/website/pages/documentation/getting_started/content/05-fundamentals/50-state-management.md [hunk 3]
```diff
@@ -82,6 +92,186 @@ def Page():
...
+Instead, you need to create a new object and assign that to a variable.
+
+```python
+a = 1
+b = a
+# a.value = 2  # ERROR: numbers are immutable
+a = 2  # instead, re-assign a new value, the number 1 will not change
+assert b == 1 # b is still 1
+```
+However, many objects in Python are mutable, including lists and dictionaries or user defined classes. This means that you can change the value of a list, dictionary, or user defined class in place without creating a new object.
```
Comment (iisakkirotko) [inline solara/website/pages/documentation/getting_started/content/05-fundamentals/50-state-management.md:107]:
```suggestion
However, many objects in Python are mutable, including lists, dictionaries, and potentially user defined classes. This means that you can change the value of a list, dictionary, or user defined class in place without creating a new object.
```

File: solara/website/pages/documentation/getting_started/content/05-fundamentals/50-state-management.md [hunk 4]
```diff
@@ -82,6 +92,186 @@ def Page():
...
+# reactive_df.value.append({"a": 4, "b": 7})  # BAD: mutating a DataFrame is not observable in Python
+df_copy = reactive_df.value.copy(deep=True)  # for Pandas 3, deep=True is not necessary
+df_copy = df_copy.append({"a": 4, "b": 7}, ignore_index=True)
+reactive_df.value = df_copy  # GOOD: re-assign a new DataFrame
+```
+
+
+## Creating a store
+
+Using reactive variables is a powerful way to manage state in your Solara applications. However, as your application grows, you may find that you need a more structured way to manage your state..
```
Comment (iisakkirotko) [inline solara/website/pages/documentation/getting_started/content/05-fundamentals/50-state-management.md:173]:
```suggestion
Using reactive variables is a powerful way to manage state in your Solara applications. However, as your application grows, you may find that you need a more structured way to manage your state.
```

**Review by iisakkirotko — CHANGES_REQUESTED** [review]

File: solara/website/pages/documentation/getting_started/content/05-fundamentals/10-components.md [hunk 5]
```diff
@@ -87,7 +87,7 @@ def MyInteractiveButton():
     solara.Button("Click me!", on_click=on_button_click)
 ```
 
-In this example, we define a function called on_button_click that will be executed when the button is clicked. In the MyInteractiveButton function, we create a Button component and set the on_click argument to the on_button_click function.
+In this example, we define a function called `on_button_click` that will be executed when the button is clicked. In the `MyInteractiveButton` function, we create a Button component and set the `on_click` argument to the `on_button_click` function.
```
[the remark this answers is not in the document]
Comment (maartenbreddels) [reply solara/website/pages/documentation/getting_started/content/05-fundamentals/10-components.md:90]: I agree, I think a follow document for more advanced component fundamentals should be added, and then we'll review this. Good point.

File: solara/website/pages/documentation/getting_started/content/05-fundamentals/50-state-management.md [hunk 6]
```diff
@@ -82,6 +92,186 @@ def Page():
...
+reactive_dict.value = {k: v for k, v in reactive_dict.value.items() if k != "a"}  # GOOD: re-assign a new dictionary
+# deleting a key (method 2)
+dict_copy = reactive_dict.value.copy()
+del dict_copy["b"]
+reactive_dict.value = dict_copy  # GOOD: re-assign a new dictionary
+```
+
+### Not mutating user defined classes
+
+Or user defined classes, like a Panda DataFrame.
```
Comment (JovanVeljanoski) [inline solara/website/pages/documentation/getting_started/content/05-fundamentals/50-state-management.md:157]:
```suggestion
Or user defined classes, like a Pandas DataFrame.
```

Comment (maartenbreddels) [reply solara/website/pages/documentation/getting_started/content/05-fundamentals/50-state-management.md]: Thanks!

**Pull request merged**